### PR TITLE
Added sequential minimum optimization algorithm for SVM

### PR DESCRIPTION
**Pull request by zhqu1148980644**
Sequential minimal optimization (SMO) is an algorithm for solving the quadratic programming (QP) problem that arises during the training of support vector machines. [wiki](https://en.wikipedia.org/wiki/Sequential_minimal_optimization)

### CODE REVIEW

**Comment by cclauss** [comment]
Could we rename the file to something more helpful like __sequential_minimum_optimization.py__ ?

File: machine_learning/sequential_minimum_optimization.py [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+    It was invented by John Platt in 1998.
+
+Input:
+    0: pandas dataframe
+    1: first column of dataframe must be tags of samples,should be 1 or -1.
+    2: rows of dataframe represent samples
+
+Usage:
+    Command:
+        python3 smo.py
```
Comment (cclauss) [inline machine_learning/sequential_minimum_optimization.py]: Change filename.

File: machine_learning/sequential_minimum_optimization.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+    Code:
+        from smo import SmoSVM, Kernel
```
Comment (cclauss) [inline machine_learning/sequential_minimum_optimization.py]: Change filename

File: machine_learning/sequential_minimum_optimization.py [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+            test_samples = self._norm(test_samples)
+
+        results = []
+        for test_sample in test_samples:
+            result = self._predict(test_sample)
+            if classify:
+                if result > 0:
+                    results.append(1)
+                else:
+                    results.append(-1)
```
Comment (cclauss) [inline machine_learning/sequential_minimum_optimization.py]: results.append(1 if result > 0 else -1)

File: machine_learning/sequential_minimum_optimization.py (hunk 1 above)
Comment (zhqu1148980644) [reply machine_learning/sequential_minimum_optimization.py]: Apologize for my carelessness.

**Comment by zhqu1148980644** [comment]
> Thank you for your pull request!🤩

Glad to be helpful.

**Comment by girisagar46** [comment]
Hi, @zhqu1148980644 

Sorry for commenting on closed PR. 🙇 

In this PR, you've mentioned this link: http://web.cs.iastate.edu/~honavar/smo-svm.pdf, but the link is not working anymore. I tried to find the replacement link for it but could not find it hence, I removed it on PR #7319 

Do you have the alternate link for it?

**Comment by zhqu1148980644** [comment]
> Hi, @zhqu1148980644
> 
> Sorry for commenting on closed PR. 🙇
> 
> In this PR, you've mentioned this link: http://web.cs.iastate.edu/~honavar/smo-svm.pdf, but the link is not working anymore. I tried to find the replacement link for it but could not find it hence, I removed it on PR #7319
> 
> Do you have the alternate link for it?

No.

**Pull request merged**